src/Features/Core/Portable/CodeFixes/AddExplicitCast/AbstractAddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,399 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeFixes.AddExplicitCast
+{
+    internal abstract class AbstractAddExplicitCastCodeFixProvider<TExpressionSyntax> : SyntaxEditorBasedCodeFixProvider
+        where TExpressionSyntax : SyntaxNode
+    {
+        /// <summary>
+        /// Give a set of least specific types with a limit, and the part exceeding the limit doesn't show any code fix,
+        /// but logs telemetry 
+        /// </summary>
+        private const int MaximumConversionOptions = 3;
+
+        internal sealed override CodeFixCategory CodeFixCategory => CodeFixCategory.Compile;
+
+        protected abstract string GetDescription(CodeFixContext context, SemanticModel semanticModel,
+            SyntaxNode? targetNode = null, ITypeSymbol? conversionType = null);
```

Comment:
1. when are targetnode or conversiontype null?  doc the cases.
2. please don't use optional params internally.  it just makes it easy to forget the values.  callers should have to think about what tehy're passing along.